File: java/third_opt/InsertMethods.java
package myz.myDBLoading;

import java.util.HashMap;

public class InsertMethods {

    // 提供id
    public static int supplyCenterID = 0;
    public static int salesmanID = 0;
    public static int productID = 0;
    public static int companyID = 0;
    public static int modelID = 0;
    public static int contractID = 0;
    public static int orderID = 0;



    //把提取数组主键内容单独做一个方法，再多加点静态变量
    public static String supplyCenterName;
    public static String salesmanNumber;
    public static String productCode;
    public static String clientEnterprise;
    public static String productModel;
    public static String contractNumber;



    public static void getPK(String[] values){
        if (values.length != 21){
            System.err.println("Wrong Array");
            return;
        }
        supplyCenterName =values[2];
        salesmanNumber =values[16];
        productCode =values[6];
        clientEnterprise=values[1];
        productModel =values[8];
        contractNumber =values[0];

    }




    //用来去重的集合
    private static HashMap<String, Integer> supplyCenters = new HashMap<>();
    public static void insertSupplyCenter(String[] line, StatementPreparer sp){

        //判断是否重复
        if (supplyCenters.containsKey(supplyCenterName)) {
             return; }

        String directorName = line[14];
        supplyCenterID++;
        //如果不重复
        //填数据
        sp.fillOneData(""+supplyCenterID, UsedDataType.INT, 1);
        sp.fillOneData(supplyCenterName, UsedDataType.STRING,2);
        sp.fillOneData(directorName,UsedDataType.STRING,3);
        //更新
        supplyCenters.put(supplyCenterName,supplyCenterID);
        sp.update();

    }


    //用来去重的集合
    private static HashMap<String, Integer> products = new HashMap<>();
    public static void insertProduct(String[] line, StatementPreparer sp){

        //判断是否重复
        if (products.containsKey(productCode)) {
            return; }

        productID++;
        String product_name = line[7];
        //如果不重复
        //填数据
        sp.fillOneData(""+productID, UsedDataType.INT, 1);
        sp.fillOneData(productCode, UsedDataType.STRING,2);
        sp.fillOneData(product_name,UsedDataType.STRING,3);
        //更新
        products.put(productCode,productID);
        sp.update();

    }


    //用来去重的集合
    private static HashMap<String, Integer> salesmanNumbers = new HashMap<>();
    public static void insertSalesman(String[] line, StatementPreparer sp){

//        String salesmanNumber =line[16];
        //判断是否重复
        if (salesmanNumbers.containsKey(salesmanNumber)) {
             return; }

        salesmanID++;
        //处理数据
        String salesmanName = line[15];
        String gender = line[17];
        switch (gender){
            case "Female":
                gender = "F";
                break;
            case "Male":
                gender = "M";
                break;
            default:
                gender = "U";
        }
        String age = line[18];
        String mobile_phone = line[19];


        //如果不重复
        //填数据
        sp.fillOneData(""+salesmanID, UsedDataType.INT, 1);
        sp.fillOneData(""+supplyCenters.get(supplyCenterName), UsedDataType.INT, 2);
        sp.fillOneData(salesmanNumber, UsedDataType.INT, 3);
        sp.fillOneData(salesmanName, UsedDataType.STRING,4);
        sp.fillOneData(gender,UsedDataType.STRING,5);
        sp.fillOneData(age, UsedDataType.INT, 6);
        sp.fillOneData(mobile_phone, UsedDataType.STRING, 7);
        //更新
        salesmanNumbers.put(salesmanNumber,salesmanID);
//        salesmanID++;
        sp.update();

    }


    //用来去重的集合
    private static HashMap<String, Integer> companys = new HashMap<>();
    public static void insertCompany(String[] line, StatementPreparer sp){

        //判断是否重复
        if (companys.containsKey(clientEnterprise)) {
            return; }

        companyID++;
        String country = line[3];
        String city = line[4];
        String industry = line[5];

        //如果不重复
        //填数据
        sp.fillOneData(""+companyID, UsedDataType.INT, 1);
        sp.fillOneData(""+supplyCenters.get(supplyCenterName), UsedDataType.INT, 2);
        sp.fillOneData(clientEnterprise, UsedDataType.STRING, 3);
        sp.fillOneData(country, UsedDataType.STRING,4);
        sp.fillOneData(city,UsedDataType.STRING,5);
        sp.fillOneData(industry,UsedDataType.STRING,6);
        //更新
        companys.put(clientEnterprise,companyID);
        sp.update();

    }

    //用来去重的集合
    private static HashMap<String, Integer> productModels = new HashMap<>();
    public static void insertProductModel(String[] line, StatementPreparer sp){

        //判断是否重复
        if (productModels.containsKey(productModel)) {
            return; }

        modelID++;
        String unitPrice = line[9];

        //如果不重复
        sp.fillOneData(""+modelID, UsedDataType.INT, 1);
        sp.fillOneData(""+products.get(productCode), UsedDataType.INT,2);
        sp.fillOneData(productModel,UsedDataType.STRING,3);
        sp.fillOneData(unitPrice,UsedDataType.INT,4);
        //更新
        productModels.put(productModel,modelID);
        sp.update();

    }

    //用来去重的集合
    private static HashMap<String, Integer> contracts = new HashMap<>();
    public static void insertContract(String[] line, StatementPreparer sp){

        //判断是否重复
        if (contracts.containsKey(contractNumber)) {
            return; }


        contractID++;
        String contractDate =line[11];

        //如果不重复
        sp.fillOneData(""+contractID, UsedDataType.INT, 1);
        sp.fillOneData(contractNumber,UsedDataType.STRING,2);
        sp.fillOneData(""+companys.get(clientEnterprise), UsedDataType.INT,3);
        sp.fillOneData(contractDate,UsedDataType.DATE,4);
        //更新
        contracts.put(contractNumber,contractID);
        sp.update();

    }



    public static void insertOrders(String[] line, StatementPreparer sp){

        String quantity = line[10];
        String estimatedDeliveryDate = line[12];
        String lodgementDate = line[13];



        //如果不重复
        sp.fillOneData(""+orderID, UsedDataType.INT, 1);
        sp.fillOneData(""+contracts.get(contractNumber), UsedDataType.INT, 2);
        sp.fillOneData(""+productModels.get(productModel), UsedDataType.INT, 3);
        sp.fillOneData(""+salesmanNumbers.get(salesmanNumber), UsedDataType.INT, 4);
        sp.fillOneData(quantity,UsedDataType.INT,5);
        sp.fillOneData(estimatedDeliveryDate, UsedDataType.DATE,6);
        sp.fillOneData(lodgementDate,UsedDataType.DATE,7);
        //更新
        orderID++;
        sp.update();

    }



    //

    //

}

File: py/InsertMethods.py
import string
from typing import List

import PrepareStatement
from DataType import DataType


class InsertMethods:
    _supplyCenterId_ = 0
    _companyId_ = 0
    _orderId_ = 0
    _contractId_ = 0
    _salesmanId_ = 0
    _productModelId_ = 0
    _productId_ = 0

    _supplyCenter_dict_ = {}
    _company_dict_ = {}
    _order_dict_ = {}
    _contract_dict_ = {}
    _salesman_dict_ = {}
    _product_dict_ = {}
    _productModel_dict_ = {}

    def __init__(self):
        self.supplyCenter = None
        self.salesmanNumber = None
        self.productCode = None
        self.companyName = None
        self.productModel = None
        self.contractNumber = None

    def setValues(self, values):
        if not len(values) == 20:
            raise RuntimeError("Invalid input")

        self.supplyCenter = values[2]
        self.salesmanNumber = values[16]
        self.productCode = values[6]
        self.companyName = values[1]
        self.productModel = values[8]
        self.contractNumber = values[0]

    def insertSupplyCenter(self, line: List[string], ps: PrepareStatement):
        if self.supplyCenter in self._supplyCenter_dict_:
            return

        self._supplyCenterId_ += 1
        self._supplyCenter_dict_[self.supplyCenter] = self._supplyCenterId_

        director = line[14]
        ps.insertData(f"{self._supplyCenterId_}", DataType.INT, 1)
        ps.insertData(self.supplyCenter, DataType.STRING, 2)
        ps.insertData(director, DataType.STRING, 3)
        ps.addIntoList()

    def insertCompany(self, line: List[string], ps: PrepareStatement):
        if self.companyName in self._company_dict_:
            return

        self._companyId_ += 1
        self._company_dict_[self.companyName] = self._companyId_

        country = line[3]
        city = line[4]
        industry = line[5]

        ps.insertData(f"{self._companyId_}", DataType.INT, 1)
        ps.insertData(f"{self._supplyCenter_dict_[self.supplyCenter]}", DataType.INT, 2)
        ps.insertData(self.companyName, DataType.STRING, 3)
        ps.insertData(country, DataType.STRING, 4)
        ps.insertData(city, DataType.STRING, 5)
        ps.insertData(industry, DataType.STRING, 6)
        ps.addIntoList()

    def insertContract(self, line: List[string], ps: PrepareStatement):
        if self.contractNumber in self._contract_dict_:
            return

        self._contractId_ += 1
        self._contract_dict_[self.contractNumber] = self._contractId_

        contractDate = line[11]

        ps.insertData(f"{self._contractId_}", DataType.INT, 1)
        ps.insertData(f"{self._company_dict_[self.companyName]}", DataType.INT, 2)
        ps.insertData(self.contractNumber, DataType.STRING, 3)
        ps.insertData(contractDate, DataType.DATE, 4)
        ps.addIntoList()

    def insertSalesman(self, line: List[string], ps: PrepareStatement):
        if self.salesmanNumber in self._salesman_dict_:
            return

        self._salesmanId_ += 1
        self._salesman_dict_[self.salesmanNumber] = self._salesmanId_

        salesmanName = line[15]
        gender = line[17]
        match gender:
            case 'Male':
                gender = 'M'

            case 'Female':
                gender = 'F'

            case 'Unknown':
                gender = 'U'

        age = line[18]
        mobile_phone = line[19]

        ps.insertData(f"{self._salesmanId_}", DataType.INT, 1)
        ps.insertData(f"{self._supplyCenter_dict_[self.supplyCenter]}", DataType.INT, 2)
        ps.insertData(self.salesmanNumber, DataType.STRING, 3)
        ps.insertData(salesmanName, DataType.STRING, 4)
        ps.insertData(gender, DataType.STRING, 5)
        ps.insertData(age, DataType.INT, 6)
        ps.insertData(mobile_phone, DataType.STRING, 7)
        ps.addIntoList()

    def insertOrder(self, line: List[string], ps: PrepareStatement):
        self._orderId_ += 1

        quantity = line[10]
        estimatedDeliveryDate = line[12]
        lodgementDate = line[13]

        ps.insertData(f"{self._orderId_}", DataType.INT, 1)
        ps.insertData(f"{self._contract_dict_[self.contractNumber]}", DataType.INT, 2)
        ps.insertData(f"{self._productModel_dict_[self.productModel]}", DataType.INT, 3)
        ps.insertData(f"{self._salesman_dict_[self.salesmanNumber]}", DataType.INT, 4)
        ps.insertData(quantity, DataType.INT, 5)
        ps.insertData(estimatedDeliveryDate, DataType.DATE, 6)
        ps.insertData(lodgementDate, DataType.DATE, 7)
        ps.addIntoList()

    def insertProductModel(self, line: List[string], ps: PrepareStatement):
        if self.productModel in self._productModel_dict_:
            return

        self._productModelId_ += 1
        self._productModel_dict_[self.productModel] = self._productModelId_

        unitPrice = line[9]

        ps.insertData(f"{self._productModelId_}", DataType.INT, 1)
        ps.insertData(f"{self._product_dict_[self.productCode]}", DataType.INT, 2)
        ps.insertData(self.productModel, DataType.STRING, 3)
        ps.insertData(unitPrice, DataType.INT, 4)
        ps.addIntoList()

    def insertProduct(self, line: List[string], ps: PrepareStatement):
        if self.productCode in self._product_dict_:
            return

        self._productId_ += 1
        self._product_dict_[self.productCode] = self._productId_

        product_name = line[7]
        ps.insertData(f"{self._productId_}", DataType.INT, 1)
        ps.insertData(self.productCode, DataType.STRING, 2)
        ps.insertData(product_name, DataType.STRING, 3)
        ps.addIntoList()
